include/deal.II/base/exception_macros.h

```diff
@@ -870,39 +870,43 @@
  * This macro is first used in step-7, where we show another example of
  * a context where it is frequently used.
  */
-#define DEAL_II_ASSERT_UNREACHABLE()                             \
-  ::dealii::deal_II_exceptions::internals::issue_error_noreturn( \
-    ::dealii::deal_II_exceptions::internals::ExceptionHandling:: \
-      abort_or_throw_on_exception,                               \
-    __FILE__,                                                    \
-    __LINE__,                                                    \
-    __PRETTY_FUNCTION__,                                         \
-    nullptr,                                                     \
-    nullptr,                                                     \
-    ::dealii::StandardExceptions::ExcMessage(                    \
-      "The program has hit a line of code that the programmer "  \
-      "marked with the macro DEAL_II_ASSERT_UNREACHABLE() to "   \
-      "indicate that the program should never reach this "       \
-      "location. You will have to find out (best done in a "     \
-      "debugger) why that happened. Typical reasons include "    \
-      "passing invalid arguments to functions (for example, if " \
-      "a function takes an 'enum' with two possible values "     \
-      "as argument, but you call the function with a third "     \
-      "value), or if the programmer of the code that triggered " \
-      "the error believed that a variable can only have "        \
-      "specific values, but either that assumption is wrong "    \
-      "or the computation of that value is buggy."               \
-      "\n\n"                                                     \
-      "In those latter conditions, where some internal "         \
-      "assumption is not satisfied, there may not be very "      \
-      "much you can do if you encounter such an exception, "     \
-      "since it indicates an error in deal.II, not in your "     \
-      "own program. If that is the situation you encounter, "    \
-      "try to come up with "                                     \
-      "the smallest possible program that still demonstrates "   \
-      "the error and contact the deal.II mailing lists with it " \
-      "to obtain help."))
-
+#if DEAL_II_KOKKOS_VERSION_GTE(3, 6, 0)
+#  define DEAL_II_ASSERT_UNREACHABLE() \
+    Kokkos::abort("ASSERT_UNREACHABLE reached.");
+#else
+#  define DEAL_II_ASSERT_UNREACHABLE()                             \
```

Comment:
Well, that's undesirable: We have an error text that spans more than 20 lines, but with new enough Kokkos versions it doesn't get shown. There ought to be a better way to achieve what you're trying to do.

Reply:
Kokkos has two macros `KOKKOS_IF_ON_HOST` and `KOKKOS_IF_ON_DEVICE` that can be used to have different codes on the CPU and the GPU. Though I don't know if we can use these macros when defining a macro.

Reply:
Agreed. The new solution should do something more reasonable.